### PR TITLE
[SPARK-47799][BUILD] Add `-g` to javac compile parameters when using SBT package jar

### PR DESCRIPTION
**Pull request by cxzl25**
### What changes were proposed in this pull request?
This PR aims to add `-g` to javac compile parameters when using SBT package jar.

> -g
Generates all debugging information, including local variables. By default, only line number and source file information is generated.

https://docs.oracle.com/en/java/javase/17/docs/specs/man/javac.html

### Why are the changes needed?
`maven-compiler-plugin` defaults to debug=true, `plexus-compiler-javac` will add the parameter `-g`.

SBT does not have this behavior by default, which leads to some differences between the jars of maven and sbt builds, although the code logic is the same.

https://maven.apache.org/plugins/maven-compiler-plugin/compile-mojo.html#debug

https://github.com/apache/maven-compiler-plugin/blob/736da68adf543cf56cd82a68e5ad28d397ace2f4/src/main/java/org/apache/maven/plugin/compiler/AbstractCompilerMojo.java#L734

https://github.com/codehaus-plexus/plexus-compiler/blob/6ae79d7f2feca3a02e75f7661468abac6a9a0a11/plexus-compilers/plexus-compiler-javac/src/main/java/org/codehaus/plexus/compiler/javac/JavacCompiler.java#L279-L285

### Does this PR introduce _any_ user-facing change?
No

### How was this patch tested?
GA

```bash
./build/sbt common-utils/package
```

local test

### Current

image

### PR
image

### Was this patch authored or co-authored using generative AI tooling?
No

### CODE REVIEW

**Review by dongjoon-hyun — COMMENTED** [review]
Thank you for making a PR, @cxzl25 .

File: project/SparkBuild.scala [hunk 1]
```diff
@@ -311,6 +311,7 @@ object SparkBuild extends PomBuild {
 
     (Compile / javacOptions) ++= Seq(
       "-encoding", UTF_8.name(),
+      "-g",
```
Comment (dongjoon-hyun) [inline project/SparkBuild.scala:314]: This looks safe to me. BTW, we still have differences, don't we?
> SBT does not have this behavior by default, which leads to some differences between the jars of maven and sbt builds, 

The file size seems to be increased from 386053 to 387076 with this, but it's still smaller than Maven-generated size, `397996`.

Comment (dongjoon-hyun) [reply project/SparkBuild.scala:314]: BTW, this is bigger than the PR title.
> Preserve parameter information when using SBT package jar

1. If this is only for parameter info, `-parameters` might be better.
2. However, to be consistent Maven, `-g` is correct. 

Given (2), I guess we need to revise the PR title, @cxzl25 .

Comment (cxzl25) [reply project/SparkBuild.scala:314]: >  we need to revise the PR title

OK

> but it's still smaller than Maven-generated size

In this PR, the main difference between the two is the META-INF folder, maven generated more information, but SBT did not.


image


image

**Review by dongjoon-hyun — APPROVED** [review]
+1, LGTM. Thank you for the fix and the explanation, @cxzl25 .

BTW, which decompiler did you use for you screenshot?

**Comment by cxzl25** [comment]
Thanks dongjoon.

> which decompiler did you use for you screenshot

I use jd-gui.
https://github.com/java-decompiler/jd-gui

**Comment by dongjoon-hyun** [comment]
Thank you for the info. :)

**Pull request closed without merging**